### PR TITLE
Async combobox

### PR DESCRIPTION
**Pull request by LFDanLu**
Closes https://github.com/adobe/react-spectrum/issues/1184

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@adobe/spectrum-css-temp/components/search/index.css [hunk 1]
```diff
@@ -74,6 +78,10 @@ governing permissions and limitations under the License.
   .spectrum-Textfield.spectrum-Search.is-valid & {
     padding-inline-end: calc(var(--spectrum-search-padding-right) + var(--spectrum-textfield-padding-x) + var(--spectrum-icon-checkmark-medium-width) + var(--spectrum-textfield-icon-margin-left));
   }
+
+  .spectrum-Textfield.spectrum-Search.is-loading & {
```
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/search/index.css:82]: I think these class names `is-invalid` `is-valid` were added recently, all of them should probably conform to the state classname pattern though
`spectrum-Search--isInvalid`
`spectrum-Search--isValid`
`spectrum-Search--isLoading`

File: packages/@adobe/spectrum-css-temp/components/textfield/index.css [hunk 2]
```diff
@@ -163,6 +163,10 @@ governing permissions and limitations under the License.
     padding-inline-end: calc(var(--spectrum-textfield-padding-x) + var(--spectrum-icon-alert-medium-width) + var(--spectrum-textfield-icon-margin-left));
   }
 
+  .spectrum-Textfield.is-loading & {
```
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/textfield/index.css:166]: same here

File: packages/@adobe/spectrum-css-temp/components/search/index.css (hunk 1 above)
Comment (snowystinger) [reply packages/@adobe/spectrum-css-temp/components/search/index.css:82]: for reference, see quiet https://github.com/adobe/react-spectrum/pull/1335/files#diff-cd526b7742dcb468383ec2ee115e3ed927d5100519b2e7d6e5a730b6968a04f9R222
`'spectrum-Textfield--quiet': isQuiet`
whether we keep the `is` in front doesn't matter to me, it's state, so it'll be read as a boolean anyways

File: packages/@react-spectrum/picker/stories/Picker.stories.tsx [hunk 3]
```diff
@@ -487,7 +487,7 @@ storiesOf('Picker', module)
   .add(
     'isLoading more',
     () => (
-      <Picker label="Test" isLoading items={flatOptions}>
+      <Picker label="Test" isLoading isOpen items={flatOptions}>
```
Comment (snowystinger) [inline packages/@react-spectrum/picker/stories/Picker.stories.tsx:490]: why is this needed

Comment (LFDanLu) [reply packages/@react-spectrum/picker/stories/Picker.stories.tsx:490]: Ah lol, forgot to remove, was testing something

File: packages/@adobe/spectrum-css-temp/components/search/index.css (hunk 1 above)
Comment (LFDanLu) [reply packages/@adobe/spectrum-css-temp/components/search/index.css:82]: Sounds good, I'll change the combobox/textfield specific patterns in this pull and open an issue to do the same for things like stepper, checkbox, etc

File: packages/@react-stately/data/src/useAsyncList.ts [hunk 4]
```diff
@@ -243,34 +272,50 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
...
   // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  // Reload on changes to filter text but only if performing server side filtering
+  // Kinda weird since it will trigger a reload when user selects a option
+  useEffect(() => {
+    if (!filterFn) {
+      dispatchFetch({type: 'loading'}, load);
+    }
+  }, [data.filterText, filterFn, load, dispatchFetch]);
```
Comment (LFDanLu) [inline packages/@react-stately/data/src/useAsyncList.ts:296]: Tried to do this outside of `useAsyncList` via calling `list.reload` in `onInputChange` but this means `filterText` is a step behind due to `setFilter` being called in the same `onInputChange`

Also kinda weird that a reload will get triggered when the user selects an option since that changes the filter text

File: packages/@react-stately/data/src/useAsyncList.ts [hunk 5]
```diff
@@ -243,34 +272,50 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
...
     },
     reload() {
       dispatchFetch({type: 'loading'}, load);
     },
     loadMore() {
-      // Ignore if already loading more.
-      if (data.state === 'loadingMore' || data.cursor == null) {
+      // Ignore if already loading or loading more.
+      // TODO: ask if this is appropriate
+      if (data.state === 'loadingMore' || data.state === 'loading' || data.cursor == null) {
```
Comment (LFDanLu) [inline packages/@react-stately/data/src/useAsyncList.ts:318]: If the user triggers a brand new filter load by typing when the dropdown is open, loadMore actually gets called by listbox (due to scroll position? not super sure will need to dig) and thus clears the item list due to the `loadingMore` -> `loading` case above. I figure we should early return anyways if `loadingMore` is triggered when `loading` is happening anyway

File: packages/@react-types/combobox/src/index.d.ts [hunk 6]
```diff
@@ -10,9 +10,10 @@
  * governing permissions and limitations under the License.
  */
 
-import {CollectionBase, DOMProps, FocusableProps, InputBase, SingleSelection, SpectrumLabelableProps, StyleProps, TextInputBase, Validation} from '@react-types/shared';
+import {AsyncLoadable, CollectionBase, DOMProps, FocusableProps, InputBase, SingleSelection, SpectrumLabelableProps, StyleProps, TextInputBase, Validation} from '@react-types/shared';
 
-export interface ComboBoxProps<T> extends CollectionBase<T>, SingleSelection, InputBase, TextInputBase, DOMProps, Validation, FocusableProps {
+// TODO: Should AsyncLoadable go in ComboBoxProps or SpectrumComboBoxProps
```
Comment (LFDanLu) [inline packages/@react-types/combobox/src/index.d.ts:15]: Open question, I figure async loading isn't spectrum specific so I put it here

File: packages/@react-stately/data/test/useAsyncList.test.js [hunk 7]
```diff
@@ -644,7 +644,7 @@ describe('useAsyncList', () => {
       console.error = consoleError;
     });
 
-    it('should throw if updating the list while loading', async () => {
+    it.skip('should throw if updating the list while loading', async () => {
```
Comment (LFDanLu) [inline packages/@react-stately/data/test/useAsyncList.test.js:647]: skipping test just so it build, but brings up the question if this test is still valid since changing the filterText is a `update` that can happen during `loading`.

File: packages/@react-stately/data/src/useAsyncList.ts [hunk 8]
```diff
@@ -243,34 +271,47 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
...
   useEffect(() => {
-    dispatchFetch({type: 'loading'}, load);
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
+    if (isInitialLoad.current || !filterFn) {
+      dispatchFetch({type: 'loading'}, load);
+    }
+    isInitialLoad.current = false;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [data.filterText, filterFn]);
```
Comment (LFDanLu) [inline packages/@react-stately/data/src/useAsyncList.ts:292]: bit concerned that I couldn't include `load` in the dep array without running into an infinite loop. Not entirely sure what causes `load` to change, something to do with the dispatches in `dispatchFetch`

Comment (LFDanLu) [reply packages/@react-stately/data/src/useAsyncList.ts:292]: maybe doesn't matter if we don't expect load to be changed by the user?

File: packages/@react-stately/data/src/useAsyncList.ts [hunk 9]
```diff
@@ -170,6 +177,15 @@ function reducer<T, C>(data: AsyncListState<T, C>, action: Action<T, C>): AsyncL
...
             abortController: action.abortController
           };
+        case 'update':
+          // TODO: Ask if this is appropriate, I want to cancel prior loads if the user is rapidly typing
+          // but "update" also covers a bunch of other actions (like item insertion)
+          // Was loading but update happened, so abort previous load.
+          data.abortController.abort();
+          return {
+            ...data,
+            ...action.updater(data)
```
[the remark this answers is not in the document]
Comment (LFDanLu) [reply packages/@react-stately/data/src/useAsyncList.ts:186]: perhaps alternative is to overwrite the `setFilterText` that comes from `useListData` and classify it as a different action than `update` so I can have this logic only happen for when filterText is being changed

**Comment by LFDanLu** [comment]
Tests to come later one some of this behavior is OK'd

File: packages/@react-spectrum/combobox/stories/ComboBox.stories.tsx [hunk 10]
```diff
@@ -411,7 +420,148 @@ storiesOf('ComboBox', module)
...
+      onInputChange={list.setFilterText}
+      isLoading={list.isLoading}
+      onLoadMore={list.loadMore}>
+      {item => <Item key={item.name}>{item.name}</Item>}
+    </ComboBox>
+  );
+}
+
+
+function AsyncLoadingExample() {
```
Comment (LFDanLu) [inline packages/@react-spectrum/combobox/stories/ComboBox.stories.tsx:536]: Note that this example will close the menu every time the combobox text is changed. IMO this feels kinda weird, perhaps I should change it so the dropdown stays open and the contents just replaced with a loading swirl

File: packages/@adobe/spectrum-css-temp/components/search/index.css [hunk 11]
```diff
@@ -67,11 +71,15 @@ governing permissions and limitations under the License.
     -webkit-appearance: none;
   }
 
-  .spectrum-Textfield.spectrum-Search.is-invalid & {
+  .spectrum-Textfield.spectrum-Search--invalid & {
```
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/search/index.css:74]: ah, I see where this (is-invalid) came from, I think spectrum-css used it as a utility class
since we have css modules though, i think the change I suggested makes more sense

**Comment by LFDanLu** [comment]
Note: currently doesn't work that well in the server async story, star wars api returns a http page cursor and `.replace` to modify the http to https doesn't quite work cuz it aint async

File: packages/@react-spectrum/textfield/src/TextFieldBase.tsx [hunk 12]
```diff
@@ -22,6 +22,8 @@ import {Label} from '@react-spectrum/label';
 import {LabelPosition} from '@react-types/shared';
 import labelStyles from '@adobe/spectrum-css-temp/components/fieldlabel/vars.css';
 import {mergeProps} from '@react-aria/utils';
+// TODO: figure out if ProgressCircle should go into ComboBox (add a prop to pass a free form element to the validation icon spot?) or stay here. Will also affect css.
+import {ProgressCircle} from '@react-spectrum/progress';
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-spectrum/textfield/src/TextFieldBase.tsx:26]: I think putting it in ComboBox makes sense for now. Perhaps we can pull a "TextField + ProgressCircle" component out to be reused later if needed, but I don't think it should be in the `textfield` package.

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-stately/data/src/useAsyncList.ts:186]: I think you may need to overwrite that function and instead trigger a `loading` action, or perhaps a new action type. You may want to use a new action type to avoid the items being removed while loading (similar to `loadingMore` and `sorting`.

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 4 above)
Comment (devongovett) [reply packages/@react-stately/data/src/useAsyncList.ts]: I think you could get rid of this if you overwrite `setFilterText` to perform a new action

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 5 above)
Comment (devongovett) [reply packages/@react-stately/data/src/useAsyncList.ts:314]: I think if you use a new action type for filtering that behaves closer to sorting/loading more and doesn't clear it might solve this

File: packages/@react-stately/data/src/useListData.ts [hunk 13]
```diff
@@ -10,16 +10,20 @@
...
   initialItems?: T[],
   /** The keys for the initially selected items. */
   initialSelectedKeys?: 'all' | Iterable<Key>,
+  /** The initial text to filter the list by. */
+  initialFilterText?: string,
   /** A function that returns a unique key for an item object. */
-  getKey?: (item: T) => Key
+  getKey?: (item: T) => Key,
+  /** A function that returns whether a item matches the current filter text. */
+  filterFn?: (item: T, filterText: string) => boolean
```
Comment (devongovett) [inline packages/@react-stately/data/src/useListData.ts:26]: Maybe just `filter`?

File: packages/@react-types/combobox/src/index.d.ts (hunk 6 above)
Comment (devongovett) [reply packages/@react-types/combobox/src/index.d.ts:15]: Interesting question. There's nothing in react-aria or stately for it so I'd say it's spectrum specific at least for now.

File: packages/@react-stately/data/src/useAsyncList.ts [hunk 14]
```diff
@@ -256,21 +307,27 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
...
+    [data.items, data.filterText, filter]);
+
   return {
-    items: data.items,
+    items: filteredItems,
     selectedKeys: data.selectedKeys,
     sortDescriptor: data.sortDescriptor,
-    isLoading: data.state === 'loading' || data.state === 'loadingMore' || data.state === 'sorting',
+    // TODO: add isFiltering? This is so we can have the loading icon in the textfield when filter text changes instead of in the listbox
+    isLoading: data.state === 'loading' || data.state === 'loadingMore' || data.state === 'sorting' || data.state === 'filtering',
```
Comment (LFDanLu) [inline packages/@react-stately/data/src/useAsyncList.ts:319]: Open question. Reason behind the ask is that the combobox needs to show the progress circle in the input field when loading (whether from filtering causing a new fetch request or just if the menu isn't open) but it also needs to show the progress circle at the bottom of the menu if a `loadMore` is triggered via scrolling. Right now I have the progress circle in the input field always show if `isLoading` is true but this leads to scenarios where you can see both the input field progress circle and the listbox progress circle (e.g. in the client side filtering story if there aren't many items in the list but it is still fetching more).

File: packages/@react-spectrum/combobox/src/MobileComboBox.tsx [hunk 15]
```diff
@@ -421,7 +463,9 @@ function ComboBoxTray(props: ComboBoxTrayProps) {
             )
           }
           ref={listBoxRef}
-          onScroll={onScroll} />
+          onScroll={onScroll}
+          onLoadMore={onLoadMore}
+          isLoading={isLoading} />
```
Comment (LFDanLu) [inline packages/@react-spectrum/combobox/src/MobileComboBox.tsx:468]: Added `isLoading` here so the client side filtering story would continue to fetch items if there weren't enough to fill the tray (if it doesn't fill the tray, then the user can't scroll to trigger loadMore)

Trying to figure out a better way to do this so that the scroll wheel only shows up when the user actually triggers a scroll event

**Comment by devongovett** [comment]
FYI I think the CSS might be a bit broken now.

image

**Comment by LFDanLu** [comment]
@devongovett Hrmmm, it looks ok to me locally but I do see the above in the built link

File: packages/@react-spectrum/combobox/src/ComboBox.tsx [hunk 16]
```diff
@@ -220,9 +261,17 @@ const ComboBoxInput = React.forwardRef(function ComboBoxInput(props: ComboBoxInp
...
+              styles,
+              'spectrum-InputGroup-input-validationIcon'
+            )
+          }
           isDisabled={isDisabled}
           isQuiet={isQuiet}
-          validationState={validationState} />
+          validationState={validationState}
+          isLoading={isLoading}
+          loadingIcon={loading} />
```
Comment (devongovett) [inline packages/@react-spectrum/combobox/src/ComboBox.tsx:274]: Should this be more like `spectrum-Textfield--loadable`? i.e. should we always reserve space for the spinner if it could show up so that when it does it doesn't shift the contents of the textfield? In that case, we could just look for the presence of the `loadingIcon` prop?

Also IMO we could rename `loadingIcon` to e.g. `loadingIndicator` since it's not really an icon?

File: packages/@react-stately/data/src/useAsyncList.ts [hunk 17]
```diff
@@ -160,11 +178,21 @@ function reducer<T, C>(data: AsyncListState<T, C>, action: Action<T, C>): AsyncL
         case 'loading':
         case 'loadingMore':
         case 'sorting':
+        case 'filtering':
+          // If there isn't a abortController provided by the action and it is a filtering action (aka clientside filtering), it is an filterText update
```
Comment (devongovett) [inline packages/@react-stately/data/src/useAsyncList.ts:182]: Why is this a separate case from the below? And why would the `abortController` not be set? I think we should probably abort the previous filtering request when the text updates.

File: packages/@react-stately/data/src/useAsyncList.ts [hunk 18]
```diff
@@ -204,6 +232,25 @@ function reducer<T, C>(data: AsyncListState<T, C>, action: Action<T, C>): AsyncL
             items: action.type === 'loading' ? [] : data.items,
             abortController: action.abortController
           };
+        case 'filtering':
```
Comment (devongovett) [inline packages/@react-stately/data/src/useAsyncList.ts:235]: Could this case be combined with the above `loading` and `sorting` cases? I think the current state value that is set above is wrong though, it should be `action.type` not always `loading`.

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 14 above)
Comment (devongovett) [reply packages/@react-stately/data/src/useAsyncList.ts:319]: Maybe we should also return the state string and rather than an `isLoading` prop on the ComboBox have a `loadingState` prop? e.g. `loadingState={list.state}` This way the ComboBox could distinguish between the different types of loads.

File: packages/@react-stately/data/src/useAsyncList.ts [hunk 19]
```diff
@@ -256,21 +307,27 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
...
       return data.items.find(item => getKey(item) === key);
     },
     reload() {
       dispatchFetch({type: 'loading'}, load);
     },
     loadMore() {
-      // Ignore if already loading more.
-      if (data.state === 'loadingMore' || data.cursor == null) {
+      // Ignore if already loading more or if performing server side filtering.
+      if (data.state === 'loadingMore' || (data.state === 'filtering' && !filter) || data.cursor == null) {
```
Comment (devongovett) [inline packages/@react-stately/data/src/useAsyncList.ts:330]: hmm what prompted this?

File: packages/@react-stately/data/src/useAsyncList.ts [hunk 20]
```diff
@@ -281,6 +338,14 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
     },
     ...createListActions({...options, getKey}, fn => {
       dispatch({type: 'update', updater: fn});
-    })
+    }),
+    setFilterText(filterText: string) {
+      if (filter && data.items.length > 0) {
+        // If client side filtering and items already exist, don't sent a fetch request, just update filterText
+        dispatch({type: 'filtering', filterText});
```
Comment (devongovett) [inline packages/@react-stately/data/src/useAsyncList.ts:345]: Would it make it any easier if this was just a `type: 'update'` as above and we reserved the `filtering` action for async loading?

File: packages/@react-spectrum/combobox/src/ComboBox.tsx (hunk 16 above)
Comment (LFDanLu) [reply packages/@react-spectrum/combobox/src/ComboBox.tsx:274]: That sounds fine, lemme adjust

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 17 above)
Comment (LFDanLu) [reply packages/@react-stately/data/src/useAsyncList.ts:182]: This is a separate case in consideration of the two kinds of "filtering" actions that can take place aka client side filtering vs server side filtering.

For client side filtering, I do a `dispatch` instead of a `dispatchFetch` and I don't include an `abortController` since client side filtering should just change the filterText in `data`. Its basically just an "update" action and thus doesn't interrupt any fetches that are currently occurring.

For server side filtering, it has an `abortController` and properly aborts previous filtering requests, acting similarly to the other actions.

I guess I could change it to be a `update` action for client side filtering, but does the above logic make sense? Lemme know if there are any flaws in that approach, greatly appreciate a second opinion

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 18 above)
Comment (LFDanLu) [reply packages/@react-stately/data/src/useAsyncList.ts:235]: Ahh, yeah I think I can combine them not sure why I didn't, good catch

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 14 above)
Comment (LFDanLu) [reply packages/@react-stately/data/src/useAsyncList.ts:319]: Lemme give that a shot

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 19 above)
Comment (LFDanLu) [reply packages/@react-stately/data/src/useAsyncList.ts:330]: This wasn't prompted by any kind of broken behavior, more just me thinking through the cases/states that could happen. I figured if server side filtering was happening (and was exceptionally slow), the user could trigger a `loadMore`  by scrolling through the existing items and hitting the bottom of the listbox. If that happens, I figured we should still wait for the server side filtering to finish rather than dispatching a new load more request (feels extraneous especially since we wouldn't have the proper up to date cursor)

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 20 above)
Comment (LFDanLu) [reply packages/@react-stately/data/src/useAsyncList.ts:345]: Yeah, I believe that would work as well. I avoided that simply because it would require adding additional `update` cases to the `loading` and `loadingMore` cases in the reducer (kinda spooky since `update` actions also happen with the other listActions above). Probably won't be too bad though, lemme try

File: packages/@react-spectrum/combobox/src/ComboBox.tsx (hunk 16 above)
Comment (LFDanLu) [reply packages/@react-spectrum/combobox/src/ComboBox.tsx:274]: Looks a little funny for the tray search field but stops the shifting that you mentioned previously:
image

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 20 above)
Comment (LFDanLu) [reply packages/@react-stately/data/src/useAsyncList.ts:345]: Ah also, are we worried about this test case: https://github.com/adobe/react-spectrum/blob/main/packages/@react-stately/data/test/useAsyncList.test.js#L647?
Seems to indicate to me that we don't want certain update actions to happen while `loading` is happening? Looking at the actions in useListData, I can see most of them being problematic if done while `loading` and `loadingMore` (except filterText heh)

File: packages/@react-types/combobox/src/index.d.ts [hunk 21]
```diff
@@ -48,12 +48,14 @@ export interface ComboBoxProps<T> extends CollectionBase<T>, SingleSelection, In
...
   /** Whether the ComboBox should be displayed with a quiet style. */
   isQuiet?: boolean,
   /**
    * Direction the menu will render relative to the ComboBox.
    * @default 'bottom'
    */
-  direction?: 'bottom' | 'top'
+  direction?: 'bottom' | 'top',
+  /** The current loading state of the ComboBox. Determines whether or not the progress circle should be shown. */
+  loadingState?: 'loading' | 'sorting' | 'loadingMore' | 'error' | 'idle' | 'filtering'
```
Comment (LFDanLu) [inline packages/@react-types/combobox/src/index.d.ts:60]: Feels odd to have all of these extra states when ComboBox only uses `loading`/`loadingMore`/`filtering`, but gotta  match the type that `state` has from `useAsyncList`. Actually might be nice to have validationState = invalid if `error` is true in `loadingState` perhaps

**Comment by bryantabaird** [comment]
Thank you for the continual review of this PR. Our team is very anxious for the useAsyncList filtering functionality that this PR introduces.

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 20 above)
Comment (devongovett) [reply packages/@react-stately/data/src/useAsyncList.ts:338]: Hmm, maybe we could simplify this a bit by only allowing using `load` for filtering in `useAsyncList` rather than also a client side filter function. Client side filtering could also be implemented within `load` so I don't really see a reason to support two different ways of doing the same thing.

File: packages/@react-types/combobox/src/index.d.ts (hunk 21 above)
Comment (devongovett) [reply packages/@react-types/combobox/src/index.d.ts:60]: I think it's ok. We may want to use more of these in the future (e.g. error states). Maybe we could move the type into a shared type?

File: packages/@react-spectrum/combobox/src/ComboBox.tsx (hunk 16 above)
Comment (devongovett) [reply packages/@react-spectrum/combobox/src/ComboBox.tsx]: Maybe we should swap the location of the clear button and the loading indicator, so the loading indicator is to the left of the clear button? And also reserve space for both of them at all times? We should probably also only show the loading indicator if there is text in the input so that it doesn't show up without the clear button...

File: packages/@react-stately/data/src/useAsyncList.ts (hunk 20 above)
Comment (LFDanLu) [reply packages/@react-stately/data/src/useAsyncList.ts:338]: Haven't tried this yet but just thinking this over: we would still need this dispatch to update the filterText right? We also wouldn't want to cancel any `loadMore` actions if filterText updated for the client side filtering case so this logic would still need exist.

File: packages/@react-spectrum/combobox/src/ComboBox.tsx (hunk 16 above)
Comment (LFDanLu) [reply packages/@react-spectrum/combobox/src/ComboBox.tsx]: Sure I can go ahead and do that. As for showing the loading indicator only if there is text in the input, are we concerned about if the field is empty and the first load takes a while? The user would only see a empty tray and no indicator of the current loading state.

Perhaps I could have the clear button appear at all times? Open to ideas

Comment (LFDanLu) [reply packages/@react-spectrum/combobox/src/ComboBox.tsx]: Another issue that comes up is the positioning of the validation icons. By flipping the loading indicator position and the clear button position, the clear button overlaps the validation icon since they are now placed in the same position. Changing the positioning of the validation icon is a bit tricky cuz it is within the textfield css

Comment (LFDanLu) [reply packages/@react-spectrum/combobox/src/ComboBox.tsx]: For now I've made it so the clear button appears at all times and have flipped the clear button's position with the loading indicator/validation icon (leveraged the `validationIconClassName` prop I forgot I added) position. Also made it so the clear button and validation icon positions for non-async comboboxes were flipped for consistency. Lemme know what you think

File: packages/@react-spectrum/combobox/src/MobileComboBox.tsx [hunk 22]
```diff
@@ -371,16 +415,19 @@ function ComboBoxTray(props: ComboBoxTrayProps) {
...
+          loadingIndicator={loadingState != null && loadingCircle}
           validationState={validationState}
-          wrapperChildren={(state.inputValue !== '' && !props.isReadOnly) && clearButton}
+          wrapperChildren={!props.isReadOnly && clearButton}
           UNSAFE_className={
             classNames(
               searchStyles,
               'spectrum-Search',
               'spectrum-Textfield',
+              'spectrum-Search--loadable',
```
Comment (LFDanLu) [inline packages/@react-spectrum/combobox/src/MobileComboBox.tsx:427]: Made it always have the loadable styles so that the validation icon and clear button for non-async comboboxes are ordered the same way for consistency (e.g. validation icon is now on the left and clear button is on the right). Note that this change is only for mobile combobox's search input and normal searchfields still have the validation icon positioned to the right of the clear button.

Comment (LFDanLu) [reply packages/@react-spectrum/combobox/src/MobileComboBox.tsx:427]: Open to naming it something different, couldn't think up of a better name

**Review by devongovett — APPROVED** [review]
